api/routes/auth.js

```
@@ -0,0 +1,111 @@
+const router = require('express').Router()
+const User = require('../models/user')
+const bcrypt = require('bcrypt')
+const jsonwebtoken = require('jsonwebtoken')
+const { sign, verify } = require('jsonwebtoken')
+const { SECRET_KEY } = process.env
+
+const generateToken = (id) => {
+  const payload = { id }
+  const options = { expiresIn: '1 day' }
+  return sign(payload, SECRET_KEY, options)
+}
+
+router.post('/signup', async (req, res, next) => {
+  const status = 201
+
+  try {
+    const { username, password } = req.body
+    const rounds = 10
+    const hashed = await bcrypt.hash(password, rounds)
+
+    const alreadyExists = await User.findOne({ username })
+    if (alreadyExists) {
+      const error = new Error(`Username '${username}' is already taken.`)
+      error.status = 400
+
+      return next(error)
+    } else {
+      const user = await User.create({ username, password: hashed })
+      const token = generateToken(user._id)
+
+      res.status(status).json({ status, user, token })
+    }
+
+  } catch (error) {
+    const e = new Error(error)
+    e.status = 400
+    next(e)
+  }
+
+})
+
+router.post('/login', async (req, res, next) => {
+  const status = 201
+
+  try {
+    const { username, password } = req.body
+
+    const user = await User.findOne({ username })
+    if (user) {
+      const valid = await bcrypt.compare(password, user.password)
+      if (valid) {
+        const status = 200
+        const response = 'You have successful logged in.'
+        const token = generateToken(user._id)
+        return res.status(status).json({ status, response, token })
+      } else {
+        const e = new Error('invalid login credentials')
+        e.status = 400
+        next(e)
+      }
```

With `try/catch`, we could do something like `throw new Error('invalid login credentials')`. This would move the error to your `.catch()` and shorten the "depth" of your code.